Enable UsesWebContentRestrictionsForFilter on macOS

szewai opened the pull request:
#### 57b87da812614c439c536ff2882a47b00927e685
<pre>
Enable UsesWebContentRestrictionsForFilter on macOS
<a href="https://rdar.apple.com/140813233">rdar://140813233</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=297290">https://bugs.webkit.org/show_bug.cgi?id=297290</a>

Reviewed by Per Arne Vollan.

Working towards removing WebFilterEvaluator from WebKit, as it is about to be deprecated. The replacement is
WCRBrowserEngineClient, and we&apos;ve already used the new framework on iOS. Most of the implementation already works,
except that on macOS, `allowURL` requires access to familycontrols daemon so the call needs to be made from network
process.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/ContentFilterUnblockHandler.h:
* Source/WebCore/platform/cocoa/ContentFilterUnblockHandlerCocoa.mm:
(WebCore::ContentFilterUnblockHandler::ContentFilterUnblockHandler):
(WebCore::ContentFilterUnblockHandler::needsNetworkProcess const):
(WebCore::ContentFilterUnblockHandler::requestUnblockAsync):
* Source/WebCore/platform/cocoa/ParentalControlsURLFilter.h:
* Source/WebCore/platform/cocoa/ParentalControlsURLFilter.mm:
(WebCore::ParentalControlsURLFilter::allowURL):
* Source/WebCore/platform/cocoa/ParentalControlsURLFilterParameters.h: Copied from Source/WebCore/platform/cocoa/ParentalControlsURLFilter.h.
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::allowEvaluatedURL):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/NetworkProcess/mac/com.apple.WebKit.NetworkProcess.sb.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::allowEvaluatedURL):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/UIProcess/WebFrameProxy.cpp:
(WebKit::WebFrameProxy::didHandleContentFilterUnblockNavigation):

Canonical link: <a href="https://commits.webkit.org/298683@main">https://commits.webkit.org/298683@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f303401d468e5234e13b70f8250212b6d4661e4c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/116187 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/35848 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/26394 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/122244 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/66746 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/8caf8129-08b8-4140-95af-22682200f41c) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/118076 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/36542 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/44436 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/88337 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/42831 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/868a18ed-cf95-48d3-b4c6-946b613377b9) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/119136 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/29156 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/104257 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/68685 "Passed tests") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/28268 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/22365 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/65925 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/108297 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/98547 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/22504 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/125393 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/114715 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/43081 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/32348 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/96999 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/43446 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/100448 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/96783 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24647 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/42059 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/19955 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/39028 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/42968 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/48560 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/143412 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/42435 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/36974 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/45770 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/44139 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/bd743a893de934777cc1a3295068ab0f48ab0578. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49286

pvollan commented on Source/WebKit/NetworkProcess/mac/com.apple.WebKit.NetworkProcess.sb.in:
```diff
@@ -879,3 +879,8 @@
 #endif
+
+#if HAVE(WEBCONTENTRESTRICTIONS) && PLATFORM(MAC)
+(allow mach-lookup
+    (global-name "com.apple.familycontrols"))
```
Perhaps we can conditionalize the access to this daemon on the feature being enabled? That does not need to land as part of this patch, and can be done as a follow-up, though.

pvollan replied:
Yes, the sandbox rule would look something like this:

```
    (allow mach-lookup
        (require-all
            (extension "com.apple.webkit.extension.mach")
            (global-name "com.apple.familycontrols")))

```

pvollan replied:
This can also be done in a follow-up :)

pvollan commented on Source/WebCore/platform/cocoa/ContentFilterUnblockHandlerCocoa.mm:
```diff
@@ -206,7 +220,7 @@
 }
 
-void ContentFilterUnblockHandler::requestUnblockAsync(DecisionHandlerFunction decisionHandler)
+void ContentFilterUnblockHandler::requestUnblockAsync(const DecisionHandlerFunction& decisionHandler)
```
Can the parameter be `DecisionHandlerFunction&& decisionHandler`?

szewai commented on Source/WebCore/platform/cocoa/ParentalControlsURLFilter.mm:
```diff
@@ -155,10 +156,21 @@ static void registerNotificationForWebContentFilterTypeChange()
 
     [wcrBrowserEngineClient allowURL:url.createNSURL().get() withCompletion:makeBlockPtr([completionHandler = WTFMove(completionHandler)](BOOL didAllow, NSError *) mutable {
         ASSERT(isMainThread());
+        RELEASE_LOG(ContentFiltering, "ParentalControlsURLFilter %" PUBLIC_LOG_STRING " the unblock request.\n", didAllow ? "allowed" : "did not allow");
```
Actually I think I could just print bool as result.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/f303401d468e5234e13b70f8250212b6d4661e4c. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49286

szewai committed:
Enable UsesWebContentRestrictionsForFilter on macOS
rdar://140813233
https://bugs.webkit.org/show_bug.cgi?id=297290

Reviewed by Per Arne Vollan.

Working towards removing WebFilterEvaluator from WebKit, as it is about to be deprecated. The replacement is
WCRBrowserEngineClient, and we've already used the new framework on iOS. Most of the implementation already works,
except that on macOS, `allowURL` requires access to familycontrols daemon so the call needs to be made from network
process.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/ContentFilterUnblockHandler.h:
* Source/WebCore/platform/cocoa/ContentFilterUnblockHandlerCocoa.mm:
(WebCore::ContentFilterUnblockHandler::ContentFilterUnblockHandler):
(WebCore::ContentFilterUnblockHandler::needsNetworkProcess const):
(WebCore::ContentFilterUnblockHandler::requestUnblockAsync):
* Source/WebCore/platform/cocoa/ParentalControlsURLFilter.h:
* Source/WebCore/platform/cocoa/ParentalControlsURLFilter.mm:
(WebCore::ParentalControlsURLFilter::allowURL):
* Source/WebCore/platform/cocoa/ParentalControlsURLFilterParameters.h: Copied from Source/WebCore/platform/cocoa/ParentalControlsURLFilter.h.
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::allowEvaluatedURL):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/NetworkProcess/mac/com.apple.WebKit.NetworkProcess.sb.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::allowEvaluatedURL):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/UIProcess/WebFrameProxy.cpp:
(WebKit::WebFrameProxy::didHandleContentFilterUnblockNavigation):

Canonical link: https://commits.webkit.org/298683@main
